keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,871 @@
+(Note that the mutating webhooks will be invoked BEFORE validating).
+
+![Webhook workflow diagram](./webhook-workflow.png)
+
+The webhook server will expose an HTTP endpoint such that to allow the API server to send AdmissionReview requests. Webhook server providers can [dynamically](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#configure-admission-webhooks-on-the-fly) configure what type of resources and what type of admission webhooks via creating CRs of type ValidatingWebhookConfiguration and/or MutatingWebhookConfiguration.
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the webhook server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. Please see the tables below for detailed information.
+
+Due to backwards compatibility concerns, the tightening will occur in three phases.
+
+1. The first phase is webhook-only, and will use [ratcheting validation](#backwards-compatibility). It will be the user's responsibility to clean up invalid objects which already existed before the webhook was enabled. Invalid objects are those which fail the new, stricter validation. The controller will not be able to automatically fix invalid objects, however it will apply a [label](#automatic-labelling-of-invalid-objects) to invalid objects so that users can easily locate them.
+2. The second phase can occur once all invalid objects are cleared from the cluster. It will be the cluster admin's responsibility to check and detect when it is safe to move to the second phase. The CRD schema validation will be tightened and the webhook will stick around to enforce immutability until immutable fields come to CRDs (Custom Resource Definition). This will be accompanied by a version change to make it clear the CRD is using different validation, however the storage version will be kept as `v1beta1` to ensure a [rollback](#rollback) is possible at phase 2.
+3. The storage version of the CRD will be changed from `v1beta1` to `v1`
```

Comment:
Should it be called "API version" rather than "storage version"?

Reply:
It's specifically referring to the version of the API that gets persisted in etcd, which is referred to as the [storage version](https://github.com/kubernetes-csi/external-snapshotter/blob/15773c346155827f6594d84ded7d9adf5da38824/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml#L182)

Reply:
See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#overview for information on what "storage version" means. 

For multi-versioned objects, there is one version which is stored in etcd (that version is the storage version), and it is converted on the fly when requested at different versions. To support rollback we need to keep the storage version at v1beta1, per the discussion [here](https://github.com/kubernetes/enhancements/pull/1902#discussion_r468725447).  Once we decide to no longer support rollback the storage version will change from v1beta1 to v1. Phase 3 occurs after GA, as v1 is the GA api.

Reply:
Got it.  Thanks.